feat(curriculum): add voting system lab

zairahira opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes https://github.com/freeCodeCamp/CurriculumExpansion/issues/655

<!-- Feel free to add any additional description of changes below this line -->

zairahira committed:
chore: add project structure

zairahira committed:
add hints

zairahira committed:
add initial tests

zairahira committed:
refine tests all passing

zairahira committed:
add more tests

zairahira committed:
cleanup

zairahira committed:
Merge remote-tracking branch 'upstream/main' into feat/lab-maps-sets

zairahira committed:
Apply suggestions from code review

Co-authored-by: Dario-DC <105294544+Dario-DC@users.noreply.github.com>

zairahira commented on curriculum/challenges/english/25-front-end-development/lab-voting-system/673b567e3ba535dda140d278.md:
````diff
@@ -0,0 +1,238 @@
+
+```js
+Poll Results:
+Option A: N votes
````
Is this better:

````md
```js
Poll Results:
OptionA: N votes
OptionB: N votes
.
.

/* 
sample output

Poll Results:
Turkey: 2 votes
Morocco: 1 votes
*/
```

````

ilenia-magoni replied:
that would be more clear for sure

zairahira committed:
Merge remote-tracking branch 'upstream/main' into feat/lab-maps-sets

moT01 commented on curriculum/challenges/english/25-front-end-development/lab-voting-system/673b567e3ba535dda140d278.md:
```diff
@@ -0,0 +1,238 @@
+
+5. In the `vote` function:
+
+   - If the `option` does not exist in the poll, the function should return the message `Option "<option>" does not exist.`;
```
```suggestion
   - If the `option` does not exist in the poll, the function should return the message `Option "<option>" does not exist.`.
```

moT01 reviewed (changes requested):
This is a lot of fun @zairahira 🎉 I left comments on a few minor typos below. One other issue I has was that the tests were interfering with me testing and debugging my code:

<details><summary>image</summary>

<img width="316" alt="Screenshot 2024-11-22 at 10 06 17 AM" src="https://github.com/user-attachments/assets/35dcdb34-838d-4364-8567-b026abe9ac99">

</details>

They are overwriting the things I add. I think we need to try and fix this. Can we do some reassignment to not overwrite the users polls/votes?

One other thing you could add is some user stories and tests like I mentioned on the bank account lab to have users add to their poll and make votes. e.g.

"You should have at least three options in your poll."
"Your poll should have at least three votes."

Not necessary to add that, but I think it makes it better.

zairahira committed:
Apply suggestions from code review

Co-authored-by: Tom <20648924+moT01@users.noreply.github.com>

zairahira committed:
update sample output

zairahira committed:
apply suggestion

Co-authored-by: Tom <20648924+moT01@users.noreply.github.com>

zairahira committed:
Merge remote-tracking branch 'upstream/main' into feat/lab-maps-sets

zairahira committed:
modify tests

zairahira commented:
@moT01,

I have modified the tests to not use `poll.clear()` every time. Maybe this was interfering/overriding your solutions.

I have also added the following tests/user stories:

- You should have at least three options in your `poll`.
- Your `poll` should have at least three votes.

Let me know if this is good to go.

Dario-DC reviewed (commented):
I still see the outcome of the last test in the console. Ideally, you should save a copy of the map, make your changes to test things, and then restore the map to how it was before the test.

Something like this:

```js
const pollCopy = new Map(poll);
try {
  poll.clear();
  // modify poll as needed
  // assert things
} finally {
  // the finally block is required for when the assert fails
  // otherwise you won't restore the map
  poll.clear();
  pollCopy.forEach((val, key) => {
    poll.set(key, val);
  });
}
```

zairahira committed:
Apply suggestions from code review

Co-authored-by: Dario-DC <105294544+Dario-DC@users.noreply.github.com>

zairahira committed:
modify tests

Dario-DC commented on client/i18n/locales/english/intro.json:
```diff
@@ -3096,7 +3096,12 @@
       "cvsw": { "title": "226", "intro": [] },
-      "cvs1": { "title": "227", "intro": [] },
+      "lab-voting-system": {
+        "title": "Build a Voting System Using Maps and Sets",
```
```suggestion
        "title": "Build a Voting System",
```

Dario-DC commented on client/src/pages/learn/full-stack-developer/lab-voting-system/index.md:
```diff
@@ -0,0 +1,9 @@
+---
+title: Introduction to the Build a Voting System Using Maps and Sets
```
```suggestion
title: Introduction to the Build a Voting System
```

Dario-DC commented on client/src/pages/learn/full-stack-developer/lab-voting-system/index.md:
```diff
@@ -0,0 +1,9 @@
+superBlock: full-stack-developer
+---
+
+## Introduction to the Build a Voting System Using Maps and Sets
```
```suggestion
## Introduction to the Build a Voting System
```

Dario-DC commented on curriculum/challenges/_meta/lab-voting-system/meta.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "Build a Voting System Using Maps and Sets",
```
```suggestion
  "name": "Build a Voting System",
```

Dario-DC commented on curriculum/challenges/_meta/lab-voting-system/meta.json:
```diff
@@ -0,0 +1,11 @@
+  "usesMultifileEditor": true,
+  "dashedName": "lab-voting-system",
+  "superBlock": "full-stack-developer",
+  "challengeOrder": [{ "id": "673b567e3ba535dda140d278", "title": "Build a Voting System Using Maps and Sets" }],
```
```suggestion
  "challengeOrder": [{ "id": "673b567e3ba535dda140d278", "title": "Build a Voting System" }],
```

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-voting-system/673b567e3ba535dda140d278.md:
```diff
@@ -0,0 +1,352 @@
+---
+id: 673b567e3ba535dda140d278
+title: Build a Voting System Using Maps and Sets
```
```suggestion
title: Build a Voting System
```

Dario-DC reviewed (commented):
In my previous review I forgot to mention that the title should be shortened. I think `Build a Voting System` is enough.

Other than that everything works well :tada:

zairahira committed:
Apply suggestions from code review

Co-authored-by: Dario-DC <105294544+Dario-DC@users.noreply.github.com>

moT01 reviewed (approved):
LGTM 🎉

mitaaios commented:
I know u don’t care but it was a person from ca I’m just quit man Sent from my iPhoneOn Dec 9, 2024, at 7:00 PM, Tom ***@***.***> wrote:﻿
Merged #57208 into main.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

mitaaios commented:
213 to now I was right Sent from my iPhoneOn Dec 9, 2024, at 7:00 PM, Tom ***@***.***> wrote:﻿
Merged #57208 into main.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

mitaaios commented:
Was it anyone u know Sent from my iPhoneOn Dec 9, 2024, at 7:00 PM, Tom ***@***.***> wrote:﻿
Merged #57208 into main.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>
